cpp/src/c_api/legacy_fa2.cpp

```diff
@@ -0,0 +1,291 @@
+            typename weight_t,
+            typename edge_type_type_t,
+            bool store_transposed,
+            bool multi_gpu>
+  void operator()()
+  {
+    if constexpr (!cugraph::is_candidate<vertex_t, edge_t, weight_t>::value) {
+      unsupported();
+    } else if constexpr (multi_gpu) {
+      unsupported();
+    } else if constexpr (!std::is_same_v<edge_t, int32_t>) {
+      unsupported();
+    } else {
+      auto graph =
+        reinterpret_cast<cugraph::graph_t<vertex_t, edge_t, false, false>*>(graph_->graph_);
+
+      auto edge_weights = reinterpret_cast<
+        cugraph::edge_property_t<cugraph::graph_view_t<vertex_t, edge_t, false, false>, weight_t>*>(
+        graph_->edge_weights_);
+
+      auto number_map = reinterpret_cast<rmm::device_uvector<vertex_t>*>(graph_->number_map_);
+
+      auto graph_view          = graph->view();
+      auto edge_partition_view = graph_view.local_edge_partition_view();
+
+      rmm::device_uvector<weight_t> tmp_weights(0, handle_.get_stream());
+      if (edge_weights == nullptr) {
+        tmp_weights.resize(edge_partition_view.indices().size(), handle_.get_stream());
+        cugraph::detail::scalar_fill(handle_, tmp_weights.data(), tmp_weights.size(), weight_t{1});
+      }
+
+      // Decompress to edgelist
+
+      auto [srcs, dsts, wgts, edge_ids, edge_types] =
+        cugraph::decompress_to_edgelist<vertex_t, edge_t, weight_t, int32_t, false, multi_gpu>(
+          handle_,
+          graph_view,
+          (edge_weights != nullptr) ? std::make_optional(edge_weights->view()) : std::nullopt,
+          std::nullopt,
+          std::nullopt,
+          (number_map != nullptr) ? std::make_optional(raft::device_span<vertex_t const>{
+                                      number_map->data(), number_map->size()})
+                                  : std::nullopt);
+
+      cugraph::legacy::GraphCOOView<vertex_t, edge_t, weight_t> legacy_coo_graph_view(
+        const_cast<vertex_t*>(srcs.data()),
+        const_cast<vertex_t*>(dsts.data()),
+        (edge_weights == nullptr) ? tmp_weights.data() : const_cast<weight_t*>(wgts->data()),
+        graph->number_of_vertices(),
+        edge_partition_view.number_of_edges());
+
+      cugraph::internals::GraphBasedDimRedCallback* callback = nullptr;
+
+      rmm::device_uvector<vertex_t> vertices(graph_view.local_vertex_partition_range_size(),
+                                             handle_.get_stream());
+      raft::copy(vertices.data(), number_map->data(), vertices.size(), handle_.get_stream());
+
+      rmm::device_uvector<float> pos(2 * (edge_partition_view.offsets().size() - 1),
+                                     handle_.get_stream());
+
+      cugraph::force_atlas2<vertex_t, edge_t, weight_t>(
+        handle_,
+        // rng_state_->rng_state_, # FIXME: Add support
+        legacy_coo_graph_view,
+        pos.data(),
+        max_iter_,
+        x_start_ != nullptr ? x_start_->as_type<float>() : nullptr,
```

Comment:
As mentioned above, this need to be sorted properly, so you'll need to make a copy and sort by vertex id after renumbering.

Reply:
After our discussion, I sort both the `x_start` and `y_start` by key where the key refers to the position of the `number_map`